dataprofiler/data_readers/csv_data.py

```diff
@@ -18,9 +20,14 @@
 class CSVData(SpreadSheetDataMixin, BaseData):
     """SpreadsheetData class to save and load spreadsheet data."""
 
-    data_type = "csv"
+    data_type: Optional[str] = "csv"
```

Comment:
I think this might be an issue with the original code. The value shouldn't be optional, it is only `None for the base class b/c it needs to be set by the derived. Hence, setting it to optional seems like it isn't true.

Maybe in the base we could make it an abstract property?

```python
import abc

class Base(abc.ABC):
    @property
    @abc.abstractmethod
    def data_type(self) -> str: ...

    @data_type.setter
    @abc.abstractmethod
    def data_type(self, value: str) -> None: ...

class CSVData(Base):
    data_type: str = "csv
```

Would something like this work?

Reply:
+1 yeah I'd prefer too that the child classes aren't `Optional[str]` type for `data_type` attribute CC @Sanketh7

Reply:
This last solution works well and I've committed that change.

Reply:
Oops. Looks like it makes a test fail.
```
=================================== FAILURES ===================================
_______________ TestBaseDataClass.test_can_apply_data_functions ________________
self = <dataprofiler.tests.data_readers.test_base_data.TestBaseDataClass testMethod=test_can_apply_data_functions>

    def test_can_apply_data_functions(self):
        class FakeDataClass:
            # matches the `data_type` value in BaseData for validating priority
            data_type = "FakeData"
    
            def func1(self):
                return "success"
    
        # initialize the data class
        data = BaseData(input_file_path="", data=FakeDataClass(), options={})
    
        # if the function exists in BaseData fail the test because the results
        # may become inaccurate.
        self.assertFalse(hasattr(BaseData, "func1"))
    
        with self.assertRaisesRegex(
            AttributeError,
            "Neither 'BaseData' nor 'FakeDataClass' " "objects have attribute 'test'",
        ):
            data.test
    
        # validate it will take BaseData attribute over the data attribute
>       self.assertIsNone(data.data_type)
E       AssertionError: 'FakeData' is not None

dataprofiler/tests/data_readers/test_base_data.py:96: AssertionError
```

Reply:
I assume we could change the test to detect an attribute error instead of checking if the field is None?

Reply:
Interesting, basically bc data_type doesn't exist in BaseData, this test is no longer validating priority.
Attribute error still wouldn't be correct bc FakeData as it.
We would need to test a property they both have. I think instead we can update the test to use `input_file_path`
```python
        class FakeDataClass:
            # matches the `data_type` value in BaseData for validating priority
            options = {"not_empty": "data"}
    
            def func1(self):
                return "success"
```
In the test we can assert the `options` is empty.

Reply:
Made the change here: https://github.com/capitalone/DataProfiler/pull/677/commits/2e850014d17df3baa1c6415ae0d2c6d76ecb893f

I didn't remove the data_type field from FakeDataClass because I wasn't sure if that would cause unintended side effects.